File: Graph/DFS.js
// javascript program to print dfs traversal from a given grpah
// this class represents a directed graph using adjacency list representation
//
class Graph
{
	// constructor 
	constructor(v)
	{
		this.V=v;
		this.adj = new Array(v);
		for (let i=0;i<v;i++)
			this.adj[i]=[];
	}
	// function to add an edge into the graph
	addEdge(v,w)
	{
		// And w to v's list.
		this.adj[v].push(w);
	}
	// A function used by DFS
	DFSUtil(v,visited)
	{
		// mark the current node as visited and print it
		visited[v]=true;
		console.log(v+"");
		// Recur for all the verices adjacent to this vertex
		for (let i of this.adj[v].values())
		{
			let n=i
			if(!visited[n])
				this.DFSUtil(n,visited);
		}
	}
	//the function to do DFS traversal . it uses recursive DFSUtil()
	DFS(v)
	{
		// mark all the verices as not visited (set as false by default in java)
		let visited = new Array(this.V);
		for (let i=0;i<this.v;i++)
			visited[i]=false;
		// call the recursive helper function to print DFS
		// traversal
		this.DFSUtil(v,visited);
	}
}
// Driver Code
g= new Graph(4);
g.addEdge(0,1);
g.addEdge(0,2);
g.addEdge(1,2);
g.addEdge(2,0);
g.addEdge(2,3);
g.addEdge(3,3);
console.log("followeing is Depth First Traversal");
g.DFS(2);
// This code is contributed by sheshbazzar byjuly 18,2024

File: Graph/DFS.py
from collections import defaultdict

class Graph:
    # constructor 
    def __init__(self):
        #default dictionary to store graph
        self.graph=defaultdict(list)
    #function to add an edge to graph
    def addEdge(self,u,v):
        self.graph[u].append(v)

    # a function used by DFS
    def DFSUtill(self,v,visted):
        # mark the current node as visted 
        #and print it 
        visted.add(v)
        print(v,end=" ")
        #Recur for all the vertices
        #adjacent to this vertex
        for neighbour in self.graph[v]:
            if neighbour not in visited:
                self.DFSUtill(neighbour,visited)
    # the function to do DFS traversal .It uses 
    # recursive DFSUtil()
    def DFS(self,v):
        # create a set to store visitied vertices
        visited=set()
        #call the recurisve helper function 
        #to print DFS traversal
        self.DFSUtil(v,visited)

    # Driver's code 
    if __name__=="__main__":
        g=Graph()
        g.addEdge(0,1)
        g.addEdge(0,2)
        g.addEdge(1,2)
        g.addEdge(2,0)
        g.addEdge(2,3)
        g.addEdge(3,3)

        g.DFS(2)

        # this code is contributed by sheshbazzar

